src/main/java/com/aws/greengrass/clientdevices/auth/iot/GreengrassClientFactory.java

```diff
@@ -0,0 +1,135 @@
+import software.amazon.awssdk.core.client.config.ClientOverrideConfiguration;
+import software.amazon.awssdk.core.retry.RetryMode;
+import software.amazon.awssdk.core.retry.RetryPolicy;
+import software.amazon.awssdk.http.apache.ApacheHttpClient;
+import software.amazon.awssdk.regions.Region;
+import software.amazon.awssdk.services.greengrassv2.GreengrassV2Client;
+import software.amazon.awssdk.services.greengrassv2.GreengrassV2ClientBuilder;
+import software.amazon.awssdk.services.greengrassv2data.GreengrassV2DataClient;
+import software.amazon.awssdk.services.greengrassv2data.GreengrassV2DataClientBuilder;
+
+import java.net.URI;
+import java.util.Optional;
+import javax.inject.Inject;
+
+/**
+ * Provides GG service clients built specifically for CDA use case.
+ * Ideally this factory should maintain the lifecycle of its clients,
+ * but for now its upto the consumer of the clients to maintain their lifecycle.
+ */
+public class GreengrassClientFactory {
+    private static final Logger logger = LogManager.getLogger(GreengrassClientFactory.class);
+    private final GreengrassServiceClientFactory greengrassServiceClientFactory;
+    private final LazyCredentialProvider lazyCredentialProvider;
+
+    /**
+     * Construct Greengrass Client Factory.
+     *
+     * @param greengrassServiceClientFactory GG Service Client Factory to retrieve the latest device-config
+     * @param lazyCredentialProvider Credential provider for the client
+     */
+    @Inject
+    public GreengrassClientFactory(GreengrassServiceClientFactory greengrassServiceClientFactory,
+                                   LazyCredentialProvider lazyCredentialProvider) {
+        this.greengrassServiceClientFactory = greengrassServiceClientFactory;
+        this.lazyCredentialProvider = lazyCredentialProvider;
+    }
+
+    /**
+     * Provides a new GG v2 client.
+     * Its upto the caller to maintain the lifecycle of the client.
+     * (TODO: This should ideally exist in Nucleus's GreengrassServiceClientFactory)
+     *
+     * @return GreengrassV2Client
+     */
+    public GreengrassV2Client getGGV2Client()  {
+        DeviceConfiguration deviceConfiguration = getDeviceConfiguration();
+        ApacheHttpClient.Builder httpClient = ClientConfigurationUtils
+                .getConfiguredClientBuilder(deviceConfiguration);
+
+        String awsRegion = Coerce.toString(deviceConfiguration.getAWSRegion());
+        GreengrassV2ClientBuilder clientBuilder = GreengrassV2Client.builder()
+                .httpClient(httpClient.build())
+                .credentialsProvider(lazyCredentialProvider)
+                .overrideConfiguration(
+                        ClientOverrideConfiguration.builder().retryPolicy(RetryMode.STANDARD).build());
+
+        if (Utils.isEmpty(awsRegion)) {
+            return clientBuilder.build();
+        }
+        clientBuilder.region(Region.of(awsRegion));
+        getEnvStage(deviceConfiguration)
+                .ifPresent(stage -> {
```

Comment:
region cannot be empty and neither can environment stage. If either of these are empty, we should not be continuing, these are critical errors